### PR TITLE
CBG-917: Unit tests for internal properties in custom conflict resolver

### PR DESCRIPTION
**Pull request by sarathkumarsivan**
Ensure unit test coverage for custom conflict resolver includes functions that read the internal properties from both localDocument and remoteDocument:

### CODE REVIEW

File: rest/replicator_test.go [hunk 1]
```diff
@@ -3070,3 +3072,275 @@ func TestBlipSyncNonUpgradableConnection(t *testing.T) {
...
+			conflictResolver: `function(conflict) {
+				var mergedDoc = new Object();
+				mergedDoc.source = "merged";
+				mergedDoc.outgoingDeleted = conflict.LocalDocument._deleted;
+				mergedDoc.incomingDeleted = conflict.RemoteDocument._deleted;
+				return mergedDoc;
+			}`,
+			expectedLocalBody: db.Body{
+				"incomingDeleted": false,
+				"outgoingDeleted": false,
```
Comment (bbrks) [inline rest/replicator_test.go]: Shouldn't this be true given the 2 input docs?

File: db/crud.go [hunk 2]
```diff
@@ -1037,10 +1037,16 @@ func (db *Database) resolveConflict(localDoc *Document, remoteDoc *Document, doc
...
+	localDocBody[BodyAttachments] = localDoc.SyncData.Attachments
+	localDocBody[BodyExpiry] = localDoc.SyncData.Expiry
+	localDocBody[BodyDeleted] = localDoc.Deleted
 
 	remoteDocBody := remoteDoc.GetDeepMutableBody()
 	remoteDocBody[BodyId] = remoteDoc.ID
 	remoteDocBody[BodyRev] = remoteDoc.RevID
+	remoteDocBody[BodyAttachments] = remoteDoc.DocAttachments
+	remoteDocBody[BodyExpiry] = remoteDoc.Expiry
+	remoteDocBody[BodyDeleted] = remoteDoc.Deleted
```
Comment (bbrks) [inline db/crud.go]: I'm not sure why remote and local are pulling attachments and expiry from different fields.

Could you explain why here in a comment if it's valid?

**Review by bbrks — CHANGES_REQUESTED** [review]

**Review by sarathkumarsivan — COMMENTED** [review]
Allow setting document expiry from conflict resolver function.

File: rest/replicator_test.go [hunk 3]
```diff
@@ -3071,3 +3072,335 @@ func TestBlipSyncNonUpgradableConnection(t *testing.T) {
...
+			}`,
+			expectedLocalBody: db.Body{
+				"localDocExp": "2021-08-24T17:08:29-07:00",
+				"source":      "merged",
+			},
+			expectedLocalRevID: createRevID(2, "1-b", db.Body{
+				"localDocExp": "2021-08-24T17:08:29-07:00",
+				"source":      "merged",
+			}),
+		},
```
Comment (bbrks) [inline rest/replicator_test.go]: This test doesn't seem to be doing anything different than `mergeReadIntlPropsLocalExpiry`.
Was the intention to set the fixed expiry value in an `_exp` field on the merged doc?

**Review by bbrks — COMMENTED** [review]
Have a question about one of the expiry test cases, but otherwise looks OK I think.

As an aside: The attachment merge test shows that the remote attachment isn't correctly being parsed (to turn the inline data into a digest), but that's good to know, and will be covered in a separate PR via CBG-929, so the test can be updated when that behaviour changes.

File: rest/replicator_test.go (hunk 3 above)
Comment (sarathkumarsivan) [reply rest/replicator_test.go]: Yeah, this was intended to set _exp on the merged doc. Refactored this in latest revision.

**Review by sarathkumarsivan — COMMENTED** [review]
Refactor mergeWriteIntlPropsExpiry test

**Review by bbrks — CHANGES_REQUESTED** [review]
As discussed earlier, push the expiry dates out beyond just a year (10+ years).
We don't want these to start failing because a document got expired one day.

**Review by bbrks — APPROVED** [review]
LGTM

**Pull request merged**